File: 103.binary-tree-zigzag-level-order-traversal.cpp
/*
 * @lc app=leetcode id=103 lang=cpp
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 *
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/description/
 *
 * algorithms
 * Medium (40.89%)
 * Total Accepted:    203.5K
 * Total Submissions: 497.6K
 * Testcase Example:  '[3,9,20,null,null,15,7]'
 *
 * Given a binary tree, return the zigzag level order traversal of its nodes'
 * values. (ie, from left to right, then right to left for the next level and
 * alternate between).
 * 
 * 
 * For example:
 * Given binary tree [3,9,20,null,null,15,7],
 * 
 * ⁠   3
 * ⁠  / \
 * ⁠ 9  20
 * ⁠   /  \
 * ⁠  15   7
 * 
 * 
 * 
 * return its zigzag level order traversal as:
 * 
 * [
 * ⁠ [3],
 * ⁠ [20,9],
 * ⁠ [15,7]
 * ]
 * 
 * 
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if(root==NULL) return result;
        queue<TreeNode*>q;
        q.push(root);
        bool lefttoright = true;

        while(!q.empty()){
            int size = q.size();
            vector<int>row(size);
            for(int i=0;i<size;i++){
                TreeNode* node = q.front();
                q.pop();

                int index = (lefttoright)?i:(size-i-1);
                row[index]= node->val;
                if(node->left) q.push(node->left);
                if(node->right)q.push(node->right);
            }
            lefttoright = !lefttoright;
            result.push_back(row);
        }




        return result; 
    }
};

File: 103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []

        lev = [root]
        ret = []
        direct = True

        while lev:
            new_lev = []
            if direct:
                ret.append([n.val for n in lev])
            else:
                ret.append([n.val for n in lev[::-1]])
        
            for node in lev:
                if node.left:
                    new_lev.append(node.left)
                if node.right:
                    new_lev.append(node.right)
            lev = new_lev
            direct = not direct
        return ret

